### PR TITLE
Fix memory leak with `ResetKernel`

### LINKED ISSUE #2519: Repeat simulation with memory leakage
Dear Nest Developers team:
I am trying to simulate the same nest model structure with different connection weight(for some local connections) repeatly, and i meet some problems for memory leakage in creation model even with nest.ResetKernel().

Then i test a easy demo like:
```
import nest
import psutil

for i in range(1000):
    nest.set_verbosity("M_WARNING")
    nest.ResetKernel()
    epop1_num = 1000
    epop1 = nest.Create('iaf_psc_alpha', epop1_num)
    ipop1_num = 1000
    ipop1 = nest.Create('iaf_psc_alpha', ipop1_num)
    nest.Connect(epop1, ipop1)
    p = psutil.virtual_memory()
    print('memory used:', float(p.used)/1024**3)
```

which causes increasing memory usage for 1.4M in every iteration like:

4.692104339599609
4.693546295166016
4.694988250732422
4.696430206298828
4.697872161865234
4.699314117431641
4.700756072998047
4.701717376708984

And I noticed in your nest pynest API documentation about nest.ResetKernel, you metion that 

>  The only exception is that dynamically loaded modules are not unloaded. This may change in a future version of NEST.

I wonder if there is any way to execute such loop of simulations, I hope to change the connection weight and repeat simulation for many times.  

I am really apreciate for your any reply.
Thanks!

Ken

### PR DESCRIPTION
**Pull request by hakonsbm**
Nodes are created with `new` in `GenericModel`:
https://github.com/nest/nest-simulator/blob/d7ef15baf1ce6b4453a10a0b9708b4e9477abc56/nestkernel/genericmodel.h#L154-L158

Then the node pointers are stored in `memory_` in `Model`. However the nodes are never deleted, which leads to a memory leak when calling `ResetKernel`. With only a few thousand nodes, calling `ResetKernel` multiple times can add up to hundreds of MB of leaked memory.

This PR puts Node pointers in smart pointers within `memory_`, so that the Nodes are automatically deleted when the model is destroyed. The explicit call to the node destructor in `NodeManager` is removed because it interferes with the deletion of the nodes at the end of `Model`s lifetime. This fixes #2519. @ChenK19 Can you check if #2512 is fixed by this as well?

### CODE REVIEW

**Review by nicolossus — APPROVED** [review]
LGTM!

**Comment by heplesser** [comment]
Is there any performance effect of replacing a normal with a smart pointer?

**Comment by hakonsbm** [comment]
@heplesser `Create` with this branch appears to be initially slightly slower (~4 % for 1e6 nodes), but faster after the first `ResetKernel` (~55 % s for 1e6 nodes).

**Comment by heplesser** [comment]
> @heplesser `Create` with this branch appears to be initially slightly slower (~4 % for 1e6 nodes), but faster after the first `ResetKernel` (~55 % s for 1e6 nodes).

Are there any effects on simulation time? I assume that smart pointers take more memory (reference counting needs to happen somewhere) and thus worry that they are cached less effectively, so that neuron access during spike delivery might become less efficient. And since neurons are created in a single place and deleted in a single place, I don't think we really need smart pointers.

I think in #2316 we just overlooked the need for destruction when we went away from the memory `pool`.  In the longer run, we should keep [this comment by @jougs](https://github.com/nest/nest-simulator/pull/2316#pullrequestreview-901821003) in mind.

File: nestkernel/model.h [hunk 1]
```diff
@@ -237,7 +237,7 @@ class Model
   /**
    * Memory for all nodes sorted by threads.
    */
-  std::vector< std::vector< Node* > > memory_;
```
Comment (heplesser) [inline nestkernel/model.h:240]: I would suggest to see if we can stick to a plain `Node*` here.

File: nestkernel/node_manager.cpp [hunk 2]
```diff
@@ -568,17 +568,7 @@ NodeManager::destruct_nodes_()
...
-    index t = kernel().vp_manager.get_thread_id();
-    SparseNodeArray::const_iterator n;
-    for ( n = local_nodes_[ t ].begin(); n != local_nodes_[ t ].end(); ++n )
-    {
-      // We call the destructor for each node excplicitly. This
-      // destroys the objects without releasing their memory. Since
-      // the Memory is owned by the Model objects, we must not call
-      // delete on the Node objects!
-      n->get_node()->~Node();
-    }
```
Comment (heplesser) [inline nestkernel/node_manager.cpp:580]: Here, we should delete nodes explicitly:
```c++
const auto tid = kernel().vp_manager.get_thread_id();
for ( auto node : local_nodes_[ tid ] )
{
  delete node;
}
local_nodes_[ tid ].clear();

**Review by heplesser — CHANGES_REQUESTED** [review]
I think we can solve this without smart pointers and the performance and memory cost they imply.

**Comment by heplesser** [comment]
I just did a little test, comparing a program filling a vector with 2^10 `*int` vs `std::shared_ptr<int>`, using g++12 and clang++14. In both cases, the shared pointer takes very close to 40B extra memory per vector element compared to a plain pointer. 

For more on the implementation in the gcc case, see https://gcc.gnu.org/onlinedocs/libstdc++/manual/memory.html#std.util.memory.shared_ptr.

**Review by heplesser — APPROVED** [review]

**Pull request merged**